core/domain/stats_services_test.py

```diff
@@ -14,20 +14,297 @@
+        self.assertEqual(
+            exploration_stats.state_stats_mapping.keys(), [
+                'Home', 'New state 2', 'End', 'New state'])
+        self.assertEqual(
+            exploration_stats.state_stats_mapping['New state'].to_dict(),
+            stats_domain.StateStats.create_default().to_dict())
+        self.assertEqual(
+            exploration_stats.state_stats_mapping['New state 2'].to_dict(),
+            stats_domain.StateStats.create_default().to_dict())
+
+        # Test renaming of states.
+        exploration.rename_state('New state 2', 'Renamed state')
+        exploration.version += 1
+        change_list = [{
+            'cmd': 'rename_state',
+            'old_state_name': 'New state 2',
+            'new_state_name': 'Renamed state'
+        }]
+        stats_services.handle_stats_creation_for_new_exp_version(
+            exploration, change_list)
+
+        exploration_stats = stats_services.get_exploration_stats_by_id(
+            exploration.id, exploration.version)
+        self.assertEqual(exploration_stats.exp_version, 3)
+        self.assertEqual(
+            exploration_stats.state_stats_mapping.keys(), [
+                'Home', 'End', 'Renamed state', 'New state'])
+
+        # Test deletion of states.
+        exploration.delete_state('New state')
+        exploration.version += 1
+        change_list = [{
+            'cmd': 'delete_state',
+            'state_name': 'New state'
+        }]
+        stats_services.handle_stats_creation_for_new_exp_version(
+            exploration, change_list)
+
+        exploration_stats = stats_services.get_exploration_stats_by_id(
+            exploration.id, exploration.version)
+        self.assertEqual(exploration_stats.exp_version, 4)
+        self.assertEqual(
+            exploration_stats.state_stats_mapping.keys(), [
+                'Home', 'Renamed state', 'End'])
+
+        # Test addition, renaming and deletion of states.
+        exploration.add_states(['New state 2'])
+        exploration.rename_state('New state 2', 'Renamed state 2')
+        exploration.delete_state('Renamed state 2')
+        exploration.version += 1
+        change_list = [{
+            'cmd': 'add_state',
+            'state_name': 'New state 2'
+        }, {
+            'cmd': 'rename_state',
+            'old_state_name': 'New state 2',
+            'new_state_name': 'Renamed state 2'
+        }, {
+            'cmd': 'delete_state',
+            'state_name': 'Renamed state 2'
+        }]
+        stats_services.handle_stats_creation_for_new_exp_version(
+            exploration, change_list)
+
+        exploration_stats = stats_services.get_exploration_stats_by_id(
+            exploration.id, exploration.version)
+        self.assertEqual(exploration_stats.exp_version, 5)
+        self.assertEqual(
+            exploration_stats.state_stats_mapping.keys(), [
+                'Home', 'End', 'Renamed state'])
+
+        # Test addition and multiple renames.
```

Comment:
Try some tricky cases, e.g.: "delete state with name A, add state with name B, rename state B to A"

Reply:
Done. The process seems more natural now. I don't think we should have any cases that might lead to an error.